Plugins/GitUIPluginInterfaces/ObjectId.cs

```diff
@@ -285,28 +285,40 @@ public override string ToString()
         /// Returns the first <paramref name="length"/> characters of the SHA-1 hash.
         /// </summary>
         /// <param name="length">The length of the returned string. Defaults to <c>8</c>.</param>
-        /// <exception cref="ArgumentOutOfRangeException"><paramref name="length"/> is less than zero, or more than 40.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="length"/> is less than one, or more than 40.</exception>
         [Pure]
         [SuppressMessage("Style", "IDE0057:Use range operator", Justification = "Performance")]
         public unsafe string ToShortString(int length = 8)
         {
-            if (length < 0)
+            if (length < 1)
             {
-                throw new ArgumentOutOfRangeException(nameof(length), length, "Cannot be less than zero.");
+                throw new ArgumentOutOfRangeException(nameof(length), length, "Cannot be less than one.");
             }
 
             if (length > Sha1CharCount)
             {
                 throw new ArgumentOutOfRangeException(nameof(length), length, $"Cannot be greater than {Sha1CharCount}.");
             }
 
-            Span<byte> buffer = stackalloc byte[_sha1ByteCount];
+            int allocSize = length <= 16 ? 8 : length <= 32 ? 16 : _sha1ByteCount;
```

Comment:
This is allocation on the stack. The added if statements will slow down and make the code more complex so I prefer to not change this.

Reply:
I agree that performance on this 'reduced' allocation is not a gain but I think you missed the point which is that, for most of the repositories where short hash is less than 8 characters, you ask for a lot less work in `Convert.ToHexString()`. Less data to convert, shorter string to allocate (8 instead of 40) and bonus if you ask for a short of size 8 because you don't have to substring.
And I'm pretty sure the added ifs are negligible compared to all that.

Reply:
Quick test seems to show that getting a hash of length of 8 is around 30% quicker (on a method that is already quick, I admit...)

Reply:
I tried to optimize ToString() with Span some time ago but did not get much improvements so I dropped it.
Use Vector operations may improve performance for at least creating the hashes, which is what matters.
ToString is mostly used in UI, the performance here makes no difference.

I will not vote against this, but to approve I want to stack alloc full size and optimize size 16 only.

Reply:
_I_ still prefer a shorter change but will not request changes

````
Span<byte> buffer = stackalloc byte[_sha1ByteCount];

// Optimize for short printouts
BinaryPrimitives.WriteUInt64BigEndian(buffer.Slice(0, 8), _i1);
            if (length > 16)
            {
                BinaryPrimitives.WriteUInt64BigEndian(buffer.Slice(8, 8), _i2);
                    BinaryPrimitives.WriteUInt32BigEndian(buffer.Slice(16, 4), _i3);
            }
```